# Analyzing Used Car Listings on eBay 

We will be working on a dataset of used cars from eBay, a classifieds section of the eBay website.

The dataset was originally scraped and uploaded to Kaggle. The version of the dataset we are working with is a sample of 50,000 data points that was prepared by Dataquest including simulating a less-cleaned version of the data.

The data dictionary provided with data is as follows:

* `dateCrawled` - When this ad was first crawled. All field-values are taken from this date.<br/>
* `name` - Name of the car.<br/>
* `seller` - Whether the seller is private or a dealer.<br/>
* `offerType` - The type of listing<br/>
* `price` - The price on the ad to sell the car.<br/>
* `abtest` - Whether the listing is included in an A/B test.<br/>
* `vehicleType` - The vehicle Type.<br/>
* `yearOfRegistration` - The year in which which year the car was first registered.<br/>
* `gearbox` - The transmission type.<br/>
* `powerPS` - The power of the car in PS.<br/>
* `model` - The car model name.<br/>
* `kilometer` - How many kilometers the car has driven.<br/>
* `monthOfRegistration` - The month in which which year the car was first registered.<br/>
* `fuelType` - What type of fuel the car uses.<br/>
* `brand` - The brand of the car.<br/>
* `notRepairedDamage` - If the car has a damage which is not yet repaired.<br/>
* `dateCreated` - The date on which the eBay listing was created.<br/>
* `nrOfPictures` - The number of pictures in the ad.<br/>
* `postalCode` - The postal code for the location of the vehicle.<br/>
* `lastSeenOnline` - When the crawler saw this ad last online.<br/>

The aim of this project is to clean the data and analyze the included used car listings.

In [138]:
import pandas as pd
import numpy as np

autos = pd.read_csv('autos.csv',encoding = 'Latin-1')

In [139]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [140]:
autos.head(5)

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

Our dataset contains 20 columns, most of which are stored as strings. There are a few columns with null values, but no columns have more than ~20% null values. There are some columns that contain dates stored as strings.

We'll start by cleaning the column names to make the data easier to work with.






### Clean Columns

In [141]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

We'll make a few changes here:

* Change the columns from camelcase to snakecase.<br/>
* Change a few wordings to more accurately describe the columns.

In [142]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen']

autos.head(3)

date_crawled                                        name  seller  \
0  2016-03-26 17:47:46            Peugeot_807_160_NAVTECH_ON_BOARD  privat   
1  2016-04-04 13:38:56  BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  privat   
2  2016-03-26 18:57:24                  Volkswagen_Golf_1.6_United  privat   

  offer_type   price  ab_test vehicle_type  registration_year    gearbox  \
0    Angebot  $5,000  control          bus               2004    manuell   
1    Angebot  $8,500  control    limousine               1997  automatik   
2    Angebot  $8,990     test    limousine               2009    manuell   

   power_ps   model   odometer  registration_month fuel_type       brand  \
0       158  andere  150,000km                   3       lpg     peugeot   
1       286     7er  150,000km                   6    benzin         bmw   
2       102    golf   70,000km                   7    benzin  volkswagen   

  unrepaired_damage           ad_created  num_photos  postal_code  \
0              nein  2016-03-26 00:00:00           0        79588   
1              nein  2016-04-04 00:00:00           0        71034   
2              nein  2016-03-26 00:00:00           0        35394   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37

### Initial Data Exploration and Cleaning

We'll start by exploring the data to find obvious areas where we can clean the data.

In [143]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-30 19:48:02  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_photos   postal_code            last_seen  
count                 50000     50000.0  50000.000000                50000  
unique                   76         NaN           NaN                39481  
top     2016-04-03 00:00:00         NaN           NaN  2016-04-07 06:17:27  
freq                   1946         NaN           NaN                    8  
mean                    NaN         0.0  50813.627300                  NaN  
std                     NaN         0.0  25779.747957                  NaN  
min                     NaN         0.0   1067.000000                  NaN  
25%                     NaN         0.0  30451.000000                  NaN  
50%                     NaN         0.0  49577.000000                  NaN  
75%                     NaN         0.0  71540.000000                  NaN  
max                     NaN         0.0  99998.000000                  NaN

Our initial observations:

* There are a number of text columns where all (or nearly all) of the values are the same:
  * `seller`
  * `offer_type`
* The `num_photos` column looks odd, we'll need to investigate this further.

In [144]:
autos['num_photos'].value_counts()

0    50000
Name: num_photos, dtype: int64

It looks like the `num_photos` column has `0` for every column. We'll drop this column, plus the other two we noted as mostly one value.

In [145]:
autos = autos.drop(['num_photos','seller','offer_type'],axis=1)

There are two columns, `price` and `odometer`, which are numeric values with extra characters being stored as text. We'll clean and convert these.

In [146]:
autos['odometer'] = (autos['odometer']
                     .str.replace(',','')
                     .str.replace('km','')
                     .astype(int)
)
autos.rename({'odometer':'odometer_km'},axis=1,inplace=True)
autos['odometer_km'].head(10)

0    150000
1    150000
2     70000
3     70000
4    150000
5    150000
6    150000
7    150000
8    150000
9    150000
Name: odometer_km, dtype: int64

In [147]:
autos['price'] = (autos['price']
                     .str.replace(',','')
                     .str.replace('$','')
                     .astype(int)
)
autos['price'].head(10)

0    5000
1    8500
2    8990
3    4350
4    1350
5    7900
6     300
7    1990
8     250
9     590
Name: price, dtype: int64

### Exploring Odometer and Price

In [148]:
autos.head(5)

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   price  ab_test vehicle_type  registration_year    gearbox  power_ps  \
0   5000  control          bus               2004    manuell       158   
1   8500  control    limousine               1997  automatik       286   
2   8990     test    limousine               2009    manuell       102   
3   4350  control   kleinwagen               2007  automatik        71   
4   1350     test        kombi               2003    manuell         0   

    model  odometer_km  registration_month fuel_type       brand  \
0  andere       150000                   3       lpg     peugeot   
1     7er       150000                   6    benzin         bmw   
2    golf        70000                   7    benzin  volkswagen   
3  fortwo        70000                   6    benzin       smart   
4   focus       150000                   7    benzin        ford   

  unrepaired_damage           ad_created  postal_code            last_seen  
0              nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1              nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2              nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3              nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4              nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50

In [149]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 17 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        50000 non-null  object
 1   name                50000 non-null  object
 2   price               50000 non-null  int64 
 3   ab_test             50000 non-null  object
 4   vehicle_type        44905 non-null  object
 5   registration_year   50000 non-null  int64 
 6   gearbox             47320 non-null  object
 7   power_ps            50000 non-null  int64 
 8   model               47242 non-null  object
 9   odometer_km         50000 non-null  int64 
 10  registration_month  50000 non-null  int64 
 11  fuel_type           45518 non-null  object
 12  brand               50000 non-null  object
 13  unrepaired_damage   40171 non-null  object
 14  ad_created          50000 non-null  object
 15  postal_code         50000 non-null  int64 
 16  last_seen           50

In [150]:
autos = autos[autos['price'].between(100,350000)]

In [151]:
autos['price'].value_counts().sort_index(ascending=False).head(15)

350000    1
345000    1
299000    1
295000    1
265000    1
259000    1
250000    1
220000    1
198000    1
197000    1
194000    1
190000    1
180000    1
175000    1
169999    1
Name: price, dtype: int64

In [152]:
autos['price'].value_counts().sort_index(ascending=True).head(15)

100    134
110      3
111      2
115      2
117      1
120     39
122      1
125      8
129      1
130     15
135      1
139      1
140      9
145      2
149      7
Name: price, dtype: int64

We went ahead and removed outliers in price which we determined would be a starting bid of 100  up to $350,000

In [153]:
autos['odometer_km'].value_counts().sort_index(ascending=True)

5000        760
10000       245
20000       757
30000       777
40000       814
50000      1009
60000      1153
70000      1214
80000      1412
90000      1733
100000     2101
125000     5037
150000    31212
Name: odometer_km, dtype: int64

We can see that the values in this field are rounded, which might indicate that sellers had to choose from pre-set options for this field. Additionally, there are more high mileage than low mileage vehicles.

### Exploring the date columns

There are a number of columns with date information:

* `date_crawled`
* `registration_month`
*`registration_year`
* `ad_created`
* `last_seen`


These are a combination of dates that were crawled, and dates with meta-information from the crawler. The non-registration dates are stored as strings.

We'll explore each of these columns to learn more about the listings.





In [15]:
autos['date_crawled'].str[:10].value_counts().sort_values(ascending=False)

2016-04-03    1862
2016-03-20    1823
2016-03-21    1794
2016-03-12    1780
2016-03-14    1768
2016-04-04    1762
2016-03-07    1739
2016-04-02    1717
2016-03-28    1686
2016-03-19    1675
2016-03-15    1655
2016-03-29    1645
2016-03-30    1627
2016-04-01    1625
2016-03-08    1600
2016-03-09    1592
2016-03-22    1586
2016-03-11    1572
2016-03-26    1558
2016-03-10    1557
2016-03-23    1557
2016-03-31    1536
2016-03-17    1519
2016-03-25    1519
2016-03-27    1501
2016-03-16    1421
2016-03-24    1420
2016-03-05    1223
2016-03-13     756
2016-03-06     677
2016-04-05     630
2016-03-18     622
2016-04-06     153
2016-04-07      67
Name: date_crawled, dtype: int64

In [16]:
autos['date_crawled'].str[:10].value_counts(normalize=True,dropna=False).sort_values(ascending=False)

2016-04-03    0.038611
2016-03-20    0.037803
2016-03-21    0.037201
2016-03-12    0.036911
2016-03-14    0.036662
2016-04-04    0.036538
2016-03-07    0.036061
2016-04-02    0.035605
2016-03-28    0.034962
2016-03-19    0.034734
2016-03-15    0.034319
2016-03-29    0.034112
2016-03-30    0.033738
2016-04-01    0.033697
2016-03-08    0.033179
2016-03-09    0.033013
2016-03-22    0.032888
2016-03-11    0.032598
2016-03-26    0.032308
2016-03-10    0.032287
2016-03-23    0.032287
2016-03-31    0.031851
2016-03-17    0.031499
2016-03-25    0.031499
2016-03-27    0.031126
2016-03-16    0.029467
2016-03-24    0.029446
2016-03-05    0.025361
2016-03-13    0.015677
2016-03-06    0.014039
2016-04-05    0.013064
2016-03-18    0.012898
2016-04-06    0.003173
2016-04-07    0.001389
Name: date_crawled, dtype: float64

Looks like the site was crawled daily over roughly a one month period in March and April 2016. The distribution of listings crawled on each day is roughly uniform.

In [17]:
autos['ad_created'].str[:10].value_counts()

2016-04-03    1874
2016-03-20    1826
2016-03-21    1805
2016-04-04    1779
2016-03-12    1772
              ... 
2016-01-29       1
2015-12-30       1
2015-09-09       1
2016-02-08       1
2016-01-22       1
Name: ad_created, Length: 76, dtype: int64

In [18]:
autos['ad_created'].str[:10].value_counts().sort_index().head(20)

2015-06-11    1
2015-08-10    1
2015-09-09    1
2015-11-10    1
2015-12-05    1
2015-12-30    1
2016-01-03    1
2016-01-07    1
2016-01-10    2
2016-01-13    1
2016-01-14    1
2016-01-16    1
2016-01-22    1
2016-01-27    3
2016-01-29    1
2016-02-01    1
2016-02-02    2
2016-02-05    2
2016-02-07    1
2016-02-08    1
Name: ad_created, dtype: int64

In [19]:
autos['ad_created'].str[:10].value_counts().sort_index(ascending=False).head(20)

2016-04-07      60
2016-04-06     157
2016-04-05     569
2016-04-04    1779
2016-04-03    1874
2016-04-02    1702
2016-04-01    1624
2016-03-31    1538
2016-03-30    1618
2016-03-29    1643
2016-03-28    1691
2016-03-27    1497
2016-03-26    1561
2016-03-25    1525
2016-03-24    1417
2016-03-23    1549
2016-03-22    1577
2016-03-21    1805
2016-03-20    1826
2016-03-19    1621
Name: ad_created, dtype: int64

In [20]:
autos['ad_created'].str[:10].value_counts().sort_values().head(15)

2016-01-22    1
2016-02-17    1
2016-02-09    1
2015-12-05    1
2015-11-10    1
2016-02-01    1
2016-02-22    1
2016-01-16    1
2016-01-14    1
2016-02-07    1
2016-01-07    1
2016-01-13    1
2016-01-03    1
2015-09-09    1
2015-12-30    1
Name: ad_created, dtype: int64

In [21]:
autos['ad_created'].str[:10].value_counts().sort_values(ascending=False).head(15)

2016-04-03    1874
2016-03-20    1826
2016-03-21    1805
2016-04-04    1779
2016-03-12    1772
2016-04-02    1702
2016-03-14    1702
2016-03-28    1691
2016-03-07    1678
2016-03-29    1643
2016-03-15    1642
2016-04-01    1624
2016-03-19    1621
2016-03-30    1618
2016-03-08    1600
Name: ad_created, dtype: int64

In [22]:
autos['ad_created'].str[:10].value_counts(normalize=True,dropna=False).sort_values(ascending=False)

2016-04-03    0.038860
2016-03-20    0.037865
2016-03-21    0.037429
2016-04-04    0.036890
2016-03-12    0.036745
                ...   
2015-11-10    0.000021
2015-12-05    0.000021
2016-02-09    0.000021
2016-02-17    0.000021
2016-01-22    0.000021
Name: ad_created, Length: 76, dtype: float64

There is a large variety of ad created dates. Most fall within 1-2 months of the listing date, but a few are quite old, with the oldest at around 9 months.

In [23]:
autos['last_seen'].str[:10].value_counts()

2016-04-06    10704
2016-04-07     6373
2016-04-05     6031
2016-03-17     1355
2016-04-03     1212
2016-04-02     1200
2016-03-30     1191
2016-04-04     1183
2016-03-31     1149
2016-03-12     1147
2016-04-01     1102
2016-03-29     1075
2016-03-22     1030
2016-03-28     1005
2016-03-20      996
2016-03-21      991
2016-03-24      953
2016-03-25      921
2016-03-23      896
2016-03-26      804
2016-03-16      793
2016-03-15      765
2016-03-19      760
2016-03-27      750
2016-03-14      609
2016-03-11      598
2016-03-10      513
2016-03-09      462
2016-03-13      428
2016-03-18      353
2016-03-08      353
2016-03-07      262
2016-03-06      208
2016-03-05       52
Name: last_seen, dtype: int64

In [24]:
autos['last_seen'].str[:10].value_counts(normalize=True,dropna=False)

2016-04-06    0.221964
2016-04-07    0.132154
2016-04-05    0.125062
2016-03-17    0.028098
2016-04-03    0.025133
2016-04-02    0.024884
2016-03-30    0.024697
2016-04-04    0.024531
2016-03-31    0.023826
2016-03-12    0.023785
2016-04-01    0.022852
2016-03-29    0.022292
2016-03-22    0.021359
2016-03-28    0.020840
2016-03-20    0.020654
2016-03-21    0.020550
2016-03-24    0.019762
2016-03-25    0.019098
2016-03-23    0.018580
2016-03-26    0.016672
2016-03-16    0.016444
2016-03-15    0.015863
2016-03-19    0.015760
2016-03-27    0.015552
2016-03-14    0.012629
2016-03-11    0.012400
2016-03-10    0.010638
2016-03-09    0.009580
2016-03-13    0.008875
2016-03-18    0.007320
2016-03-08    0.007320
2016-03-07    0.005433
2016-03-06    0.004313
2016-03-05    0.001078
Name: last_seen, dtype: float64

The crawler recorded the date it last saw any listing, which allows us to determine on what day a listing was removed, presumably because the car was sold.

The last three days contain a disproportionate amount of 'last seen' values. Given that these are 6-10x the values from the previous days, it's unlikely that there was a massive spike in sales, and more likely that these values are to do with the crawling period ending and don't indicate car sales.

In [25]:
autos['registration_year'].describe()

count    48224.000000
mean      2004.730964
std         87.897388
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The year that the car was first registered will likely indicate the age of the car. Looking at this column, we note some odd values. The minimum value is `1000`, long before cars were invented and the maximum is `9999`, many years into the future.

### Dealing with Incorrect Registration Year Data
Because a car can't be first registered before the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

One option is to remove the listings with these values. Let's determine what percentage of our data has invalid values in this column:

In [154]:
autos['registration_year'].value_counts().sort_index()

1000    1
1001    1
1111    1
1800    2
1910    2
       ..
5911    1
6200    1
8888    1
9000    1
9999    3
Name: registration_year, Length: 95, dtype: int64

In [27]:
autos['registration_year'].value_counts().sort_index().head(20)

1000    1
1001    1
1111    1
1800    2
1910    2
1927    1
1929    1
1931    1
1934    2
1937    4
1938    1
1939    1
1941    2
1943    1
1948    1
1950    1
1951    2
1952    1
1953    1
1954    2
Name: registration_year, dtype: int64

In [28]:
autos['registration_year'].value_counts().sort_index(ascending=False).head(20)

9999       3
9000       1
8888       1
6200       1
5911       1
5000       3
4800       1
4500       1
4100       1
2800       1
2019       2
2018     468
2017    1383
2016    1202
2015     380
2014     662
2013     801
2012    1308
2011    1618
2010    1587
Name: registration_year, dtype: int64

In [29]:
autos['registration_year'].value_counts(normalize=True,dropna=False).sort_index(ascending=False).head(40)

9999    0.000062
9000    0.000021
8888    0.000021
6200    0.000021
5911    0.000021
5000    0.000062
4800    0.000021
4500    0.000021
4100    0.000021
2800    0.000021
2019    0.000041
2018    0.009705
2017    0.028679
2016    0.024925
2015    0.007880
2014    0.013728
2013    0.016610
2012    0.027123
2011    0.033552
2010    0.032909
2009    0.043132
2008    0.045828
2007    0.047134
2006    0.055325
2005    0.060364
2004    0.055968
2003    0.055844
2002    0.051364
2001    0.054516
2000    0.064366
1999    0.059701
1998    0.048524
1997    0.039918
1996    0.028098
1995    0.024739
1994    0.012981
1993    0.008709
1992    0.007610
1991    0.007009
1990    0.006885
Name: registration_year, dtype: float64

In [30]:
(~autos['registration_year'].between(1900,2016)).sum() / autos.shape[0]

0.03881884538818845

Given that this is less than 4% of our data, we will remove these rows.

In [31]:
autos = autos[autos['registration_year'].between(1900,2016)]

In [32]:
autos['registration_year'].value_counts(normalize=True,dropna=False).head(10)

2000    0.066966
2005    0.062802
1999    0.062112
2004    0.058228
2003    0.058099
2006    0.057560
2001    0.056718
2002    0.053439
1998    0.050483
2007    0.049038
Name: registration_year, dtype: float64

It appears that most of the vehicles were first registered in the past 20 years.

### Exploring Price by Brand


In [34]:
autos['brand'].describe()

count          46352
unique            40
top       volkswagen
freq            9799
Name: brand, dtype: object

In [35]:
autos['brand'].value_counts()

volkswagen        9799
bmw               5107
opel              4971
mercedes_benz     4480
audi              4022
ford              3237
renault           2182
peugeot           1384
fiat              1187
seat               846
skoda              761
nissan             711
mazda              706
smart              658
citroen            651
toyota             593
hyundai            464
sonstige_autos     442
volvo              423
mini               408
mitsubishi         379
honda              365
kia                328
alfa_romeo         309
porsche            279
suzuki             275
chevrolet          263
chrysler           163
dacia              123
daihatsu           116
jeep               106
subaru              98
land_rover          98
saab                77
jaguar              71
daewoo              69
trabant             63
rover               62
lancia              49
lada                27
Name: brand, dtype: int64

In [36]:
autos['brand'].value_counts(normalize=True)

volkswagen        0.211404
bmw               0.110179
opel              0.107245
mercedes_benz     0.096652
audi              0.086771
ford              0.069835
renault           0.047075
peugeot           0.029858
fiat              0.025608
seat              0.018252
skoda             0.016418
nissan            0.015339
mazda             0.015231
smart             0.014196
citroen           0.014045
toyota            0.012793
hyundai           0.010010
sonstige_autos    0.009536
volvo             0.009126
mini              0.008802
mitsubishi        0.008177
honda             0.007875
kia               0.007076
alfa_romeo        0.006666
porsche           0.006019
suzuki            0.005933
chevrolet         0.005674
chrysler          0.003517
dacia             0.002654
daihatsu          0.002503
jeep              0.002287
subaru            0.002114
land_rover        0.002114
saab              0.001661
jaguar            0.001532
daewoo            0.001489
trabant           0.001359
r

German manufacturers represent four out of the top five brands, almost 50% of the overall listings. Volkswagen is by far the most popular brand, with approximately double the cars for sale of the next two brands combined.

There are lots of brands that don't have a significant percentage of listings, so we will limit our analysis to brands representing more than 5% of total listings.

In [53]:
brands = autos['brand'].value_counts(normalize=True)
common_brands = brands[brands > .05]
print(common_brands)


volkswagen       0.211404
bmw              0.110179
opel             0.107245
mercedes_benz    0.096652
audi             0.086771
ford             0.069835
Name: brand, dtype: float64


In [54]:
common_brands.index

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')

In [113]:
brand_mean_prices = {}

for i in common_brands.index:
    brand_only = autos[autos['brand'] == i]
    mean_price = brand_only['price'].mean()
    brand_mean_prices[i] = int(mean_price)
    
print(brand_mean_prices)

{'volkswagen': 5436, 'bmw': 8381, 'opel': 3005, 'mercedes_benz': 8672, 'audi': 9380, 'ford': 3779}


In [121]:
mean_prices = pd.Series(brand_mean_prices).sort_values(ascending=False)
print(mean_prices)

audi             9380
mercedes_benz    8672
bmw              8381
volkswagen       5436
ford             3779
opel             3005
dtype: int64


Of the top 5 brands, there is a distinct price gap:

* Audi, BMW and Mercedes Benz are more expensive
* Ford and Opel are less expensive
* Volkswagen is in between - this may explain its popularity, it may be a 'best of 'both worlds' option.

In [122]:
brand_mean_mileage = {}

for i in common_brands.index:
    brand_only = autos[autos['brand'] == i]
    mean_mileage = brand_only['odometer_km'].mean()
    brand_mean_mileage[i] = int(mean_mileage)
    
print(brand_mean_mileage)

{'volkswagen': 128799, 'bmw': 132695, 'opel': 129384, 'mercedes_benz': 131025, 'audi': 129245, 'ford': 124277}


In [123]:
mean_mileage = pd.Series(brand_mean_mileage).sort_values(ascending=False)
print(mean_mileage)

bmw              132695
mercedes_benz    131025
opel             129384
audi             129245
volkswagen       128799
ford             124277
dtype: int64


In [124]:
brand_mean_price_miles = pd.DataFrame(mean_mileage, columns=['mean_mileage'])

In [125]:
brand_mean_price_miles['mean_price'] = mean_prices

In [128]:
brand_mean_price_miles

mean_mileage  mean_price
bmw                  132695        8381
mercedes_benz        131025        8672
opel                 129384        3005
audi                 129245        9380
volkswagen           128799        5436
ford                 124277        3779

The range of car mileages does not vary as much as the prices do by brand, instead all falling within 10% for the top brands. There is a slight trend to the more expensive vehicles having higher mileage, with the less expensive vehicles having lower mileage.